core/blockchain.go

```diff
@@ -191,9 +192,11 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *par
 	if bc.genesisBlock == nil {
 		return nil, ErrNoGenesis
 	}
+	bc.hashHistory = NewHashStorage(bc.genesisBlock.Header())
```

Comment:
Just to play devils advocate: Did you consider adding a `blockhashCache` similar to what's been done for other construct above? Of course, it's not as precise as the circular buffer, but does offload the implementation/code maintenance complexity of having a custom storage code.

Reply:
Not sure I follow? The block hash and header hash is the same thing... ?

Reply:
If I understand you correctly: your point here is that headers are already cached within `HeaderChain` so that if merely caching was enough then this code would not have the substantial improvement it does. Makes sense.

Reply:
I just didn't understand what you meant with your comment about devils advocate. Could you clarify?

Reply:
Ah, I think I see now. And yes, they're already cached. Iterating from a given reference is not cheap though

Reply:
Yeah, you got it.